pimcore/models/Tool/CustomReport/Config.php

```diff
@@ -150,7 +150,23 @@ public static function getReportsList()
     public static function getAdapter($configuration, $fullConfig = null)
     {
         $type = $configuration->type ? ucfirst($configuration->type) : 'Sql';
-        $adapter = "\\Pimcore\\Model\\Tool\\CustomReport\\Adapter\\{$type}";
+        if ($type === 'Custom') {
+            if (@\class_exists($configuration->className, true)) { // can need autoloader, can send warning and break json response so we use @
+                $adapter = $configuration->className;
+                $adpaterInstance = new $adapter($configuration, $fullConfig);
+                // we make some check for security reason (because className can be choosen at the GUI level)
+                if (get_parent_class($adpaterInstance) !== 'Pimcore\\Model\\Tool\\CustomReport\\Adapter\\AbstractAdapter') {
+                    throw new \Exception("Class" . $configuration->className ." doesn't extend AbstractAdapter");
+                }
+                if ((!method_exists($adpaterInstance, 'getData')) || (!method_exists($adpaterInstance, 'getColumns')) || (!method_exists($adpaterInstance, 'getAvailableOptions'))) {
```

Comment:
\Pimcore\Model\Tool\CustomReport\Adapter\AbstractAdapter::getData is an abstract method if the instance of the class extends that Adapter, in my opinion we shouldn't check if method exists again.